usr/lib/api/api_interface.c

```diff
@@ -2752,12 +2752,11 @@ CK_RV C_Initialize(CK_VOID_PTR pVoid)
                 goto error;
             }
         }
-        // If we EVER need to create threads from this library we must
-        // check the Flags for the Can_Create_OS_Threads flag
-        // Right now the library DOES NOT create threads and therefore this
-        // check is irrelavant.
+
         if (pArg->flags & CKF_LIBRARY_CANT_CREATE_OS_THREADS) {
-            TRACE_DEVEL("Can't create OS threads...This is OK\n");
+            TRACE_ERROR("Flag CKF_LIBRARY_CANT_CREATE_OS_THREADS is set\n");
+            rc = CKR_NEED_TO_CREATE_THREADS;
```

Comment:
That is a breaking user-visible change.  Couldn't we just deactivate event support in this case without changes to the current behaviour?

Reply:
Hm, well I would actually not want to disable event support, because it is meant to be there for a good reason. At a later time we will add support for concurrent master key change which relies on the event support being available. If some applications disable event support then its extremely hard to detect this situation for an admin who changes the MKs and expects this to happen concurrently.

I see 2 options:
1.) accept that this change may cause applications for fail.
2.) ignore the CKF_LIBRARY_CANT_CREATE_OS_THREADS flag (i.e. drop this commit), since the thread really is handled totally inside OCK. This event thread will actually never be seen by any application that uses the PKCS#11 interface of OCK. Its a purely internal thing, not exposed to the outside. 

@p-steuer What's your opinion on this?

Reply:
Not allowing the lib to create threads is not the default and i cant think of any reason to do so, that means that
(1) only very few applications will set this flag
(2) those applications that do, will have a good reason to do so

My conclusion is, that ignoring the flag is not a viable option because we aim at standard compliance and because of (2).

So the options that we have are basically
(A) what is currently implemented: fail at init if flag is set
(B) what @juergenchrist suggested: disable event support in case the flag is set

(B) would basically double our test matrix because we had to test OCK with and w/o event support (just think about transactional mem vs locking builds..).

Because of that and because of (1), i would actually favor (A) i.e., going with whats implemented right now, even if it may break some applications. (Hopefully maintainers of such apps would speak up and we could understand their needs and how to adress them.)

Reply:
> Could openCrypoki ever guarantee its callers that its callees dont spawn threads ?

Well, up to now OpenCryptoki did not ITSELF spawn threads.  An application was/is free to spawn threads as it likes (and the locking stuff we added prevented OCK from failing if they do so).

Reply:
My point was that openCryptoki can not guarantee the (shared) libs it loads do not spawn threads. Only openCryptoki itself dos not spawn threads, but imo thats not enough to accept init params which have the "no threads" flag set.

Reply:
Yes, agreed. 

I would also favor (A).

Also because if there would be processes with disabled event support, we could never ever implement a reliable concurrent master key change support in the future. If some applications may not receive such events (because of LIBRARY_CANT_CREATE_OS_THREADS being set) , we never know when it is save to make the final master key switchover. Unfortunately you can't tell as pkcsslotd, nor as a process sending an admin event if a process runs without event support. 

So being able to disabling event support would be a really bad choice, making the whole event support effort basically useless....

So if we agree on (A), then I just leave that commit included and wait for further review feedback.

Reply:
I am not against (A).  Just wanted to state that _if_ an application currently sets CKF_LIBRARY_CANT_CREATE_OS_THREADS, everything works since we currently just ignore this flag.  If such an application updates OCK to this new version (which does not seem to be a major version change!), said application will no longer work.  So the effect warrants a new major version (since not backward compatible).

I can see a number of scenarios of managed (Java, .NET, ...) that want to set this flag to not have unmanaged threads outside of the virtual machine.  Yet I do not know whether or not these systems set this flag.